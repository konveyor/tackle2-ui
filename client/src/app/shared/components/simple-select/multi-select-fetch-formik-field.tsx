import React from "react";
import { FieldHookConfig, useField } from "formik";

import {
  ISimpleSelectFetchProps,
  SimpleSelectFetch,
} from "./simple-select-fetch";

export interface IMultiSelectFetchFormikFieldProps {
  fieldConfig: FieldHookConfig<any>;
  selectConfig: Omit<ISimpleSelectFetchProps, "value" | "onChange" | "onClear">;
  isEqual: (a: any, b: any) => boolean;
}

export const MultiSelectFetchFormikField: React.FC<
  IMultiSelectFetchFormikFieldProps
> = ({ fieldConfig, selectConfig, isEqual }) => {
  const [field, , helpers] = useField(fieldConfig);

  return (
    <SimpleSelectFetch
      value={field.value}
      onChange={(selection) => {
        const currentValue = field.value || [];

        let nextValue: any[];
        const elementExists = currentValue.find((f: any) => {
          return isEqual(f, selection);
        });

        if (elementExists) {
          nextValue = currentValue.filter((f: any) => !isEqual(f, selection));
        } else {
          nextValue = [...currentValue, selection];
        }

        helpers.setValue(nextValue);
      }}
      onClear={() => helpers.setValue(undefined)}
      {...selectConfig}
    />
  );
};
