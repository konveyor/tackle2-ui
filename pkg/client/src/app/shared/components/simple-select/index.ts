export * from "./simple-select";
export * from "./simple-select-fetch";
export * from "./single-select-fetch-formik-field";
export * from "./multi-select-fetch-formik-field";
export * from "./single-select-formik-field";
export * from "./single-select-fetch-optionvalue-formik-field";
export * from "./multi-select-fetch-optionvalue-formik-field";
export * from "./single-select-optionvalue-formik-field";
