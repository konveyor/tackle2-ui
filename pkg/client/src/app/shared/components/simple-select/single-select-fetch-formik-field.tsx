import React from "react";
import { FieldHookConfig, useField } from "formik";

import {
  ISimpleSelectFetchProps,
  SimpleSelectFetch,
} from "./simple-select-fetch";

export interface ISingleSelectFetchFormikFieldProps {
  fieldConfig: FieldHookConfig<any>;
  selectConfig: Omit<ISimpleSelectFetchProps, "value" | "onChange" | "onClear">;
}

export const SingleSelectFetchFormikField: React.FC<
  ISingleSelectFetchFormikFieldProps
> = ({ fieldConfig, selectConfig }) => {
  const [field, , helpers] = useField(fieldConfig);

  return (
    <SimpleSelectFetch
      value={field.value}
      onChange={(selection) => helpers.setValue(selection)}
      onClear={() => helpers.setValue(undefined)}
      {...selectConfig}
    />
  );
};
