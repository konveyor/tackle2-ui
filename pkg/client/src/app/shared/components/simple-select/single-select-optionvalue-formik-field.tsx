import React from "react";
import { FieldHookConfig, useField } from "formik";

import {
  ISimpleSelectProps,
  OptionWithValue,
  SimpleSelect,
} from "./simple-select";

export interface ISingleSelectOptionValueFormikFieldProps<T> {
  fieldConfig: FieldHookConfig<T | null>;
  selectConfig: Omit<
    ISimpleSelectProps,
    "value" | "options" | "onChange" | "onClear"
  >;
  options: T[];
  toOptionWithValue: (option: T) => OptionWithValue<T>;
}

export const SingleSelectOptionValueFormikField = <T extends any>({
  fieldConfig,
  selectConfig,
  options,
  toOptionWithValue,
}: ISingleSelectOptionValueFormikFieldProps<T>) => {
  const [field, , helpers] = useField(fieldConfig);

  return (
    <SimpleSelect
      value={field.value ? toOptionWithValue(field.value) : undefined}
      options={options.map(toOptionWithValue)}
      onChange={(selection) => {
        const selectionValue = (selection as OptionWithValue<T>).value;
        helpers.setValue(selectionValue);
      }}
      onClear={() => helpers.setValue(null)}
      {...selectConfig}
    />
  );
};
