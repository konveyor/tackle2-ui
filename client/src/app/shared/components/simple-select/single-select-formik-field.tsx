import React from "react";
import { FieldHookConfig, useField } from "formik";

import { ISimpleSelectProps, SimpleSelect } from "./simple-select";

export interface ISingleSelectFormikFieldProps {
  fieldConfig: FieldHookConfig<any>;
  selectConfig: Omit<ISimpleSelectProps, "value" | "onChange" | "onClear">;
}

export const SingleSelectFormikField: React.FC<
  ISingleSelectFormikFieldProps
> = ({ fieldConfig, selectConfig }) => {
  const [field, , helpers] = useField(fieldConfig);

  return (
    <SimpleSelect
      value={field.value}
      onChange={(selection) => helpers.setValue(selection)}
      onClear={() => helpers.setValue(undefined)}
      {...selectConfig}
    />
  );
};
