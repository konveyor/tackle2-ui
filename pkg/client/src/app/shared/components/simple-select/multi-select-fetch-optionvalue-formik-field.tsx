import React from "react";
import { FieldHookConfig, useField } from "formik";

import {
  ISimpleSelectFetchProps,
  SimpleSelectFetch,
} from "./simple-select-fetch";
import { OptionWithValue } from "./simple-select";

export interface IMultiSelectFetchOptionValueFormikFieldProps<T> {
  fieldConfig: FieldHookConfig<T[]>;
  selectConfig: Omit<
    ISimpleSelectFetchProps,
    "value" | "options" | "onChange" | "onClear"
  >;
  options: T[];
  toOptionWithValue: (option: T) => OptionWithValue<T>;
  isEqual: (a: T, b: T) => boolean;
}

export const MultiSelectFetchOptionValueFormikField = <T extends any>({
  fieldConfig,
  selectConfig,
  options,
  toOptionWithValue,
  isEqual,
}: IMultiSelectFetchOptionValueFormikFieldProps<T>) => {
  const [field, , helpers] = useField(fieldConfig);

  return (
    <SimpleSelectFetch
      value={field.value.map(toOptionWithValue)}
      options={options.map(toOptionWithValue)}
      onChange={(selection) => {
        const selectionValue = (selection as OptionWithValue<T>).value;

        const currentValue = field.value;

        let nextValue: T[];
        const elementExists = currentValue.find((f: T) => {
          return isEqual(f, selectionValue);
        });

        if (elementExists) {
          nextValue = currentValue.filter(
            (f: T) => !isEqual(f, selectionValue)
          );
        } else {
          nextValue = [...currentValue, selectionValue];
        }

        helpers.setValue(nextValue);
      }}
      onClear={() => helpers.setValue([])}
      {...selectConfig}
    />
  );
};
