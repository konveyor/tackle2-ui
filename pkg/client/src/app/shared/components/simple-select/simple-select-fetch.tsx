import React from "react";
import { AxiosError } from "axios";

import { WarningTriangleIcon } from "@patternfly/react-icons/dist/esm/icons/warning-triangle-icon";

import { getAxiosErrorMessage } from "@app/utils/utils";
import { ISimpleSelectProps, SimpleSelect } from "./simple-select";

export interface ISimpleSelectFetchProps
  extends Omit<ISimpleSelectProps, "toggleIcon" | "customContent"> {
  isFetching: boolean;
  fetchError?: AxiosError;
}

export const SimpleSelectFetch: React.FC<ISimpleSelectFetchProps> = ({
  isFetching,
  fetchError,
  ...props
}) => {
  let customContent;
  if (isFetching) {
    customContent = <div>&nbsp;Loading...</div>;
  } else if (fetchError) {
    customContent = <div>&nbsp;{getAxiosErrorMessage(fetchError)}</div>;
  }

  return (
    <SimpleSelect
      toggleIcon={fetchError && <WarningTriangleIcon />}
      customContent={customContent}
      {...props}
    />
  );
};
