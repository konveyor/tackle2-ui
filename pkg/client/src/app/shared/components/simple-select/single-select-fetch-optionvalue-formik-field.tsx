import React from "react";
import { FieldHookConfig, useField } from "formik";

import {
  ISimpleSelectFetchProps,
  SimpleSelectFetch,
} from "./simple-select-fetch";
import { OptionWithValue } from "./simple-select";

export interface ISingleSelectFetchOptionValueFormikFieldProps<T> {
  fieldConfig: FieldHookConfig<T | null>;
  selectConfig: Omit<
    ISimpleSelectFetchProps,
    "value" | "options" | "onChange" | "onClear"
  >;
  options: T[];
  toOptionWithValue: (option: T) => OptionWithValue<T>;
}

export const SingleSelectFetchOptionValueFormikField = <T extends any>({
  fieldConfig,
  selectConfig,
  options,
  toOptionWithValue,
}: ISingleSelectFetchOptionValueFormikFieldProps<T>) => {
  const [field, , helpers] = useField(fieldConfig);

  return (
    <SimpleSelectFetch
      value={field.value ? toOptionWithValue(field.value) : undefined}
      options={options.map(toOptionWithValue)}
      onChange={(selection) => {
        const selectionValue = (selection as OptionWithValue<T>).value;
        helpers.setValue(selectionValue);
      }}
      onClear={() => helpers.setValue(null)}
      {...selectConfig}
    />
  );
};
