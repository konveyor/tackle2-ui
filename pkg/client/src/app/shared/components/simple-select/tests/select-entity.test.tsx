import React from "react";
import { shallow } from "enzyme";

import { SimpleSelect } from "../simple-select";

describe("SimpleSelect", () => {
  it("Renders without crashing", () => {
    const wrapper = shallow(
      <SimpleSelect
        aria-label="my-selection"
        onChange={jest.fn()}
        options={["uno", "dos"]}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
