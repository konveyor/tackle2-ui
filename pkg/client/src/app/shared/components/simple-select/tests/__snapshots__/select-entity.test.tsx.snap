// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SimpleSelect Renders without crashing 1`] = `
<Select
  aria-describedby=""
  aria-invalid={false}
  aria-label="my-selection"
  aria-labelledby=""
  chipGroupComponent={null}
  className=""
  clearSelectionsAriaLabel="Clear all"
  createText="Create"
  customBadgeText={null}
  customContent={null}
  direction="down"
  favorites={Array []}
  favoritesLabel="Favorites"
  hasInlineFilter={false}
  inlineFilterPlaceholderText={null}
  inputIdPrefix=""
  isCreatable={false}
  isDisabled={false}
  isGrouped={false}
  isInputValuePersisted={false}
  isOpen={false}
  isPlain={false}
  menuAppendTo="inline"
  noResultsFoundText="No results found"
  onClear={[Function]}
  onCreateOption={[Function]}
  onFilter={null}
  onSelect={[Function]}
  onToggle={[Function]}
  onTypeaheadInputChanged={null}
  ouiaSafe={true}
  placeholderText="Select..."
  removeSelectionAriaLabel="Remove"
  selections={Array []}
  toggleAriaLabel="Options menu"
  toggleIcon={null}
  toggleId={null}
  typeAheadAriaLabel=""
  validated="default"
  variant="single"
  width=""
>
  <SelectOption
    className=""
    component="button"
    index={0}
    inputId=""
    isChecked={false}
    isDisabled={false}
    isFavorite={null}
    isLastOptionBeforeFooter={[Function]}
    isLoad={false}
    isLoading={false}
    isNoResultsOption={false}
    isPlaceholder={false}
    isSelected={false}
    key="0-uno"
    keyHandler={[Function]}
    onClick={[Function]}
    sendRef={[Function]}
    setViewMoreNextIndex={[Function]}
    value="uno"
  />
  <SelectOption
    className=""
    component="button"
    index={0}
    inputId=""
    isChecked={false}
    isDisabled={false}
    isFavorite={null}
    isLastOptionBeforeFooter={[Function]}
    isLoad={false}
    isLoading={false}
    isNoResultsOption={false}
    isPlaceholder={false}
    isSelected={false}
    key="1-dos"
    keyHandler={[Function]}
    onClick={[Function]}
    sendRef={[Function]}
    setViewMoreNextIndex={[Function]}
    value="dos"
  />
</Select>
`;
